### PR TITLE
Add new test case to integration tests

### PR DESCRIPTION
**Pull request by dwhjames**
Test the Amazon certificate authorities
https://www.amazontrust.com/repository/
with and without CRLs.

### CODE REVIEW

**Comment by djc** [comment]
Why?

**Comment by dwhjames** [comment]
> Why?

@djc, I initially did this privately as I was testing a use case that involves AWS and revocation checking. But I thought the extra test coverage may be of use to the project, so I’ve tidied it up to offer it back.

Due to dependency tree restrictions, I originally had to work against v0.101.7. And at that version it seemed the integration test examples for CRLs were limited in number as well as simple (single CRL); moreover, the types were more ‘awkward’. It seems like `main` is now ‘cleaner’ for handling `webpki::CertRevocationList`.

I’m not offended if this is overkill for the amount of test coverage that this might gain relative to amount of new test code: just trying to offer what might be of use (more coverage / more complex integration example) rather than keep it private.

**Comment by djc** [comment]
How much code would you have to duplicate to put these tests in a separate file? Would be interesting to look at where the extra coverage is coming from. I think it might be interesting to do this, but it also feels like overkill a bit.

How much runtime does it add in practice? Probably not a lot?

**Comment by cpu** [comment]
Thanks for sharing some tests :+1: 

> more coverage

From [codecov's perspective](https://app.codecov.io/gh/rustls/webpki/commit/70147e699d48e0c84e9bc514f1dcd0e8d1ebdae1/indirect-changes?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr%20comments&utm_term=rustls) it doesn't seem like this branch added coverage outside of a 0.07% indirect coverage improvement in [src/time.rs](https://github.com/rustls/webpki/blob/84b1463bdbe6bf03a5fdd0a0e1c7c3de2e3af702/src/time.rs#L113).

I buy that there could be interesting properties that differ between this test corpus and the existing coverage that are meaningful outside of naive coverage metrics, but I'd like to see those specific differences highlighted. I think it would be a stronger contribution if the new tests were minimized to focus on what's new vs existing.

File: tests/integration.rs [hunk 1]
```diff
@@ -19,6 +19,339 @@ use core::time::Duration;
...
+                    &intermediates_legacy,
+                    time,
+                    KeyUsage::server_auth(),
+                    crls.map(|l| revocation_options_for_test(l)),
+                    None,
+                )
+                .unwrap();
+
+            // verify should find shortest path
+            assert!(anchors.contains(path.anchor()));
```
Comment (dwhjames) [inline tests/integration.rs:287]: @cpu @djc 
I’ve scanned the codebase and this is the only test usage of `verify_for_usage` where the computed path is checked, and specifically this check establishes that the shortest path is found. With the trust anchors and intermediates given here, there are two valid paths for every leaf cert tested.

File: tests/integration.rs [hunk 2]
```diff
@@ -19,6 +19,339 @@ use core::time::Duration;
...
+                cert.verify_for_usage(
+                    webpki::ALL_VERIFICATION_ALGS,
+                    &anchors,
+                    &intermediates,
+                    time,
+                    KeyUsage::server_auth(),
+                    Some(revocation_options_for_test(crls)),
+                    None,
+                )
+                .is_err_and(|e| matches!(e, webpki::Error::CertRevoked))
```
Comment (dwhjames) [inline tests/integration.rs:326]: @cpu @djc to the best of my ability, I couldn’t see any other test cases where revocation checking occurs,
1. with real-world CRLs (as opposed to those constructed by generate.py), i.e. production data and not trivially small data (hypothesis being that real data can be weird and thus useful for finding edge cases and regressions)
2. where the collection of CRLs are from multiple authorities

Comment (dwhjames) [reply tests/integration.rs:326]: The assertion immediately above this one is complementary (shows the revoked certificates will verify if the crls from the intermediate CAs are not given).

File: tests/integration.rs [hunk 3]
```diff
@@ -19,6 +19,339 @@ use core::time::Duration;
...
+                webpki::ALL_VERIFICATION_ALGS,
+                &anchors,
+                &intermediates,
+                time,
+                KeyUsage::server_auth(),
+                None,
+                None,
+            )
+            .is_err_and(|e| matches!(e, webpki::Error::CertExpired { .. }))
+        );
```
Comment (dwhjames) [inline tests/integration.rs:351]: @cpu @djc if you’re looking for code to trim, this is likely lower value; I included this for completeness (test everything that https://www.amazontrust.com/repository/ has to offer) but there are already tests for expiry.

**Comment by dwhjames** [comment]
> How much runtime does it add in practice? Probably not a lot?

@djc, sub 1 second on my machine (2.3 GHz Intel Core i9)

```
$ cargo test --test integration --no-default-features --features alloc,std,aws-lc-rs -- --exact amazon
…
running 1 test
test amazon ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 14 filtered out; finished in 0.13s
```

File: tests/integration.rs (hunk 1 above)
Comment (cpu) [reply tests/integration.rs:287]: >  specifically this check establishes that the shortest path is found. With the trust anchors and intermediates given here, there are two valid paths for every leaf cert tested.

I believe the [better TLS](https://bettertls.com/) test coverage in [tests/better_tls.rs](https://github.com/rustls/webpki/blob/main/tests/better_tls.rs) has a variety of test cases that exercise path building like this. Did you take a peek at those?

**Comment by dwhjames** [comment]
Some additional coverage to note: these certificates cover RSA (2048, 4096) and EC (P-256, P-384) along with SHA-256 and SHA-384. I sampled around and saw some but not all of these (no RSA-4096 or EC P-384 from my admittedly incomplete sampling).

**Comment by dwhjames** [comment]
> How much code would you have to duplicate to put these tests in a separate file?

@djc, almost nothing, I think.

File: tests/integration.rs (hunk 1 above)
Comment (dwhjames) [reply tests/integration.rs:287]: I didn’t, but now I’m looking
- https://github.com/Netflix/bettertls/tree/master/test-suites/pathbuilding

I’m actually not seeing a test case that matches what is seen in figures 1 & 2 in
- https://aws.amazon.com/blogs/security/acm-will-no-longer-cross-sign-certificates-with-starfield-class-2-starting-august-2024/

of course, there are a lot of complex/pathological in that bettertls.

Note the integration of this crate with better_tls has no revocation checking.

https://github.com/rustls/webpki/blob/84b1463bdbe6bf03a5fdd0a0e1c7c3de2e3af702/tests/better_tls.rs#L79-L86

Comment (cpu) [reply tests/integration.rs:287]: > Note the integration of this crate with better_tls also has no revocation checking.

Fair point 👍

**Comment by dwhjames** [comment]
Moving test code to a separate source file in https://github.com/rustls/webpki/pull/324/commits/da11edf54cff41ac32a62c6331a3b5c3a2433588. I was interpreted this as your implied preference @djc .

**Comment by dwhjames** [comment]
@cpu is there something broken with the CI configuration?

All the integration tests appears to be filtered out, rather than running.
- https://github.com/rustls/webpki/actions/runs/13487872127/job/37681489585?pr=324

and that seems to be the case on at least the previous PR
- https://github.com/rustls/webpki/actions/runs/13458105815/job/37606542735?pr=323

**Comment by ctz** [comment]
> @cpu is there something broken with the CI configuration?

Excellent spot! Addressing this in #325

**Review by ctz — APPROVED** [review]

**Review by djc — APPROVED** [review]
